feat: added a Command-Prompt for SuperFile specific actions

Rocco-Gossmann opened the pull request:
# Description

This PR adds a Command-Prompt, similar to that of TMUX or Vim/NeoVim.
That way, we can define new functions, without having to define new shortcut for each.

The Prompt is triggered via `>`  (or "Shift + ." on QWERTZ

In Future, You could extend the Prompt to also trigger existing functionality, without having to rely on the shortcuts all the time. That way the user has more freedom, in what keys they want to map.

# Current Features / Available Commands
## implemented commands for now
- `split` => opens a new filepanel with the current filepanels location
- `cd` => change the active filepanels location to the entered path
- `open` => open a new filepanel for the given path
- `$` (dollar) => execute the input as a shell-command (same as using the "shift + ;" shortcut and entering the command there)

  
# Demo
https://github.com/user-attachments/assets/0dd7dcb1-3b1e-4a7d-9f3c-53c1aa4bfc2c


# Command Implementation
The Available Commands are Implemented in a map, with the following schema:
```go
           //  \/--------- what the user actually enters.
promptCommands["$"] = PromptCommand{
    renderPrefix: "$ ",  //<- not very relevant, but may become more important later
    
    renderHint:   "Bash/Powershell - Command", //<- Info shown in suggestions list
    
    // Function, that is executed, if the command is run
    handleCommand: func(input string, _ *promptModal, m *model) bool {

	m.openCommandLine()
	m.commandLine.input.SetValue(input)
	m.enterCommandLine()

        // Return defines, if the Prompt should close after execution
	return true

    },
}
```

# Relations
The `open` - command could solve this wish / feature-request here.
https://github.com/yorukot/superfile/issues/443#issue-2632838804

The `split`- command solves my personal gripe with the the software right here
https://github.com/yorukot/superfile/pull/732#issue-2938618992

# Disclaimer:
When it comes to TUI - Design or working with this bubbletea freamwork, I have zero experience. So if there is something horribly wrong, let me know and please tell me, how to do it better. Thank you 🙏🏼

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an enhanced prompt modal for executing commands with improved feedback and suggestions.
  - Added a new hotkey (">") to quickly access the prompt modal.
  - Incorporated a refreshed terminal icon for a clearer visual cue in the prompt modal header.

- **Refactor**
  - Updated file panel creation to allow custom starting locations.
  - Streamlined user interactions by replacing the legacy command line interface with the prompt modal.

- **Style**
  - Refined the application window title for more consistent branding.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lazysegtree commented:
Additional comment. You also need to create a website change. Put a separate PR to `1.2.2-website-changes` branch for that. 
See example PR - https://github.com/yorukot/superfile/pull/729

lazysegtree commented:
@yorukot Please do an initial feature review and provide any high level comments. Meanwhile we will finalise the implementation.

lazysegtree commented on src/internal/key_function.go:
```diff
@@ -206,6 +211,24 @@ func (m *model) typingModalOpenKey(msg string) {
+
+}
+
+func (m *model) promptModalOpenKey(msg string) {
```
Self note : This functions is to send input to promptModal. And keypress event while PromptModal is open would get consumed by this. 

Please add this todo comment : 
```
// Todo : There is a lot of duplication for these models, each one of them has to handle 
// ConfirmTyping and CancleTyping in a similar way. There is a scope of some good refactoring here.  
```

lazysegtree commented:
@Rocco-Gossmann Would you be able to right a testcase in out python testsuite for this ?

See testsuite directory for existing tests

You can have a testcase that open spf prompt and sends a cd command, and then verify that cd worked by performing and verifying a copy paste or delete or some file operation in new directory. 
The testsuite is a bit hacky, so feel free to improve it for your testcase.

lazysegtree commented on src/internal/prompt_modal.go:
```diff
@@ -0,0 +1,221 @@
+	p.open = true
+}
+
+func (p *promptModal) Close(m *model) {
```
You dont need to pass around model pointer here. 
That principle, I forgot its technical term - "Pass minimum required info to a function"

Rocco-Gossmann replied:
done

lazysegtree commented:
Haven't gone through the entire code yet. Will complete the review soon.

lazysegtree commented on src/internal/model.go:
```diff
@@ -387,6 +393,13 @@ func (m model) View() string {
 	}
 
+	if m.promptModal.open {
+		helpMenu := m.promptModalRender()
```
rename `helpMenu` to `promptModal` or something

Rocco-Gossmann replied:
done

lazysegtree commented on src/internal/model_render.go:
```diff
@@ -509,6 +509,12 @@ func (m *model) warnModalRender() string {
 	return modalBorderStyle(modalHeight, modalWidth).Render(title + "\n\n" + content + "\n\n" + tip)
 }
 
+func (m *model) promptModalRender() string {
```
Remove this function.
Replace `m.promptModalRender()` with `m.promptModal.Render(m)`

Rocco-Gossmann replied:
done

lazysegtree commented on src/internal/prompt_modal.go:
```diff
@@ -0,0 +1,221 @@
+	return p.cmd.handleCommand(p.textInput.Value(), p, m)
+}
+
+func (p *promptModal) Render(m *model) string {
```
take `renderWidth` as a paramter. This function doesn't needs entire model structure.

Rocco-Gossmann commented:
I have Some time this evening, so I'll see if i can knock of some of your points.

> Some notes
>     * We want to make **some** of these work on spf prompt
>       * `cd ~`
>       * `cd $HOME`
>       * `cd $(<a shell command by user>)`
>       * pwd - print current panels directory
>     * "not a supershell command" - This doesnt' feels correct. Should be "not a superfile prompt command"
>     * cursor is not blinking while typing at prompt. Needs to be fixed.
>     * Would like to have this covered by units tests too.

Regarding your notes though.
This is not a fully blown Programming Language (like Bash).
It is Just a simple command runner.
- So stuff like `cd $(<cmd>)` would be pretty hard to implement.
  - I'm not qualified to create my own programming language or rebuild bash in go.

- `pwd` makes no sense if you ask me. 
  - you can see what folder you are in already, through the highlighted filePanel
  
- `cd ~` and `cd $HOME` we could combine into a `home` command.
  - To cd to the users HomeDir.
  - anything else would require either interacting with the shell, which is difficult, due to windows or a fully blown parser/compiler for handling commands.
  That's unfortunately out of my league though.

Rocco-Gossmann commented on src/internal/config_type.go:
```diff
@@ -150,8 +150,10 @@ type ConfigType struct {
-	Quit    []string `toml:"quit"`
+	Confirm    []string `toml:"confirm" comment:"=================================================================================================\nGlobal hotkeys (cannot conflict with other hotkeys)"`
+	Quit       []string `toml:"quit"`
+	OpenPrompt []string `toml:"prompt"`
```
both done

Rocco-Gossmann commented on src/internal/handle_panel_navigation.go:
```diff
@@ -54,13 +54,17 @@ func (m *model) pinnedDirectory() {
 }
 
 // Create new file panel
+
```
done

Rocco-Gossmann commented on src/internal/handle_panel_navigation.go:
```diff
@@ -54,13 +54,17 @@ func (m *model) pinnedDirectory() {
 // Create new file panel
+
 func (m *model) createNewFilePanel() {
+	m.createNewFilePanelForLocation(variable.HomeDir)
```
done

Rocco-Gossmann commented:
> @Rocco-Gossmann Would you be able to write a testcase in our python testsuite for this ?
> 
> See testsuite directory for existing tests
> 
> You can have a testcase that open spf prompt and sends a cd command, and then verify that cd worked by performing and verifying a copy paste or delete or some file operation in new directory. The testsuite is a bit hacky, so feel free to improve it for your testcase.

Unfortunately I don't know how to write Python at all. Sorry.

lazysegtree commented:
Okay. I will co author this feature with you.

lazysegtree commented:
I also need to look at prompt feature of nvim, and how other file managers like yazi, lf, midnight commander does this.

Rocco-Gossmann commented:
> Additional comment. You also need to create a website change. Put a separate PR to `1.2.2-website-changes` branch for that. See example PR - #729

done
https://github.com/yorukot/superfile/pull/740

lazysegtree commented on src/internal/default_config.go:
```diff
@@ -60,6 +60,10 @@ func defaultModelConfig(toggleDotFileBool bool, toggleFooter bool, firstFilePane
 			data:        getHelpMenuData(),
 			open:        false,
 		},
+		promptModal: promptModal{
```
Please dont name variables exactly as struct typename

lazysegtree replied:
This will anyways be flagged by our linter, we are soon gonna add.

Rocco-Gossmann replied:
done [f9159da](https://github.com/yorukot/superfile/pull/736/commits/f9159da9cc6ee329386657b792222b4b694b822e)

lazysegtree commented on src/internal/prompt_modal.go:
```diff
@@ -0,0 +1,220 @@
+	PROMPT_COMMAND_SPLIT        PromptCommandPrefix = "split"
+)
+
+var promptCommands map[PromptCommandPrefix]PromptCommand
```
This map doesn't needs to be and shouldn't be a global variable. The available commands are the property of promptModal.

Rocco-Gossmann replied:
done
[740bdbc](https://github.com/yorukot/superfile/pull/736/commits/740bdbc5e252a96c1ecf72bf0eb3d83f54156f95)

lazysegtree commented:
I do plan to 
- implement parsing `$(<cmd>)` and `${ENV_VAR}`
- And add some fixes 
- Add unit testcases for promptModal
- Add testsuite test for promptModal

I will be working on this. But for now, I am focussing on a releasing 1.2.1 and merging a few PRs which are finalized, so give me some time to get back on this.

yorukot commented:
I will delete the original command input.
And make it work inside the prompt_modal.

lazysegtree commented:
FYI, I think we would merge this crash fix PR first.  
- https://github.com/yorukot/superfile/pull/730

lazysegtree commented:
<img width="1144" alt="image" src="https://github.com/user-attachments/assets/4ea3f7cc-a846-42f5-8fbf-3a1308f9e340" />

(1) Superfile file prompt should appear in the middle of the whole terminal (veritically)

(2) If we are replacing command line with spf prompt, why change hotkey from `:` to `>` . We should let it stay `:` then.

Rocco-Gossmann commented:
> (1) Superfile file prompt should appear in the middle of the whole terminal (veritically)
> 
> (2) If we are replacing command line with spf prompt, why change hotkey from `:` to `>` . We should let it stay `:` then.

Sorry, I'm not familiar with the Framework you are using here. I don't know how to change the position. I just copied the "HelpModal" 

IDK how to make the cursor blick either. Could you help me there?

lazysegtree commented:
Rebased with latest main.

Okay @Rocco-Gossmann I will fix these.

lazysegtree commented:
@Rocco-Gossmann @yorukot How about this
User would press `:` to open the prompt.
Then he can press > to run spf commands, or start typing directly to execute normal shell commands

- `:` then  `mkdir hi`
- `:` then `> split`
- `:` then `> cd ${MY_ENV_VAR}`

lazysegtree commented on src/internal/prompt_modal.go:
```diff
@@ -0,0 +1,242 @@
+type PromptCommand struct {
+	renderPrefix  string
+	renderHint    string
+	handleCommand func(input string, p *promptModalType, m *model) bool
```
This doesn't looks right. PromptCommand is a lower level struct. It should never be aware of and use high level objects like `model` directly. 

This is breaking encapsulation.

I will fix it.

lazysegtree replied:
Same goes with `PromptCommand` being aware of `promptModalType`.

Rocco-Gossmann replied:
> This doesn't looks right. PromptCommand is a lower level struct. It should never be aware of and use high level objects like `model` directly.
> 
> This is breaking encapsulation.
> 
> I will fix it.

I put that in there, because all commands need to have access to the model, to control it.

Rocco-Gossmann replied:
> Same goes with `PromptCommand` being aware of `promptModalType`.

`p *promptModalType` is there, so, that you can use it to display error-messages.

yorukot commented:
@lazysegtree 
Yeah sure,that would be a great way to implement also have a pretty good user experience.

Rocco-Gossmann commented:
> @Rocco-Gossmann @yorukot How about this User would press `:` to open the prompt. Then he can press > to run spf commands, or start typing directly to execute normal shell commands
> 
> * `:` then  `mkdir hi`
> * `:` then `> split`
> * `:` then `> cd ${MY_ENV_VAR}`

I find the other way around less confusing.
By saying. "The default mode of this is the `shell-command` mode", you make the user think that the spf-commands are part of the shell.

While having the spf-command mode first´. you are telling the user "this prompt controls spf and doing shell-commands is a spf-function"

lazysegtree commented:
@coderabbitai pause

lazysegtree commented:
@Rocco-Gossmann The PR is getting bloated with comments. Could you join discord ?

lazysegtree commented:
About default mode being shell command mode. 
its because thats what is most used. 

The most used path should be easiest and most straightforward to access.

Rocco-Gossmann commented:
> @Rocco-Gossmann The PR is getting bloated with comments. Could you join discord ? We will continue our discussion there. Or lets create a discussion on discussion tab.

@lazysegtree @yorukot 
If you don't mind, I think I'll drop out at this point. 

1.) I don't like Discord. Don't have an account, Don't want one. 

2.) You clearly want this PR to be more, than I ever intended it to be.

3.) Just for the record. I created this PR after this comment by @lazysegtree https://github.com/yorukot/superfile/pull/733#issuecomment-2745181292

The idea was to NOT confuse the user between Command-Prompt and Shell-Commandline.
Now you are walking back on that and making the Command-Prompt into the Shell-Commandline.
Triggering the actual Command-Prompt functions via a Prefix.

At this point, we could have just keept the original Shell-Commandline and just replaced the `spf`-prefix (I implemented here: https://github.com/yorukot/superfile/pull/733) with something else.

lazysegtree commented:
Moved discussion here - https://github.com/yorukot/superfile/discussions/745

lazysegtree commented:
Merging this to a non-main branch. 
Will start a new PR, as this PR has too many comments that are not relevant now.
